# KUIS 2 Pembelajaran Mesin

    Nama    : Adika Ahmad hanif Nazhir  
    Kelas   : TI - 3B  
    NIM     : 2041720171

## Tentang dataset

Pada sebuah lembaga kemanusiaan berbasis teknologi informasi, dibentuk beberapa tim kemanusiaan yang bertugas untuk terjun ke masyarakat memberikan bantuan dalam bentuk aplikasi komputer atau layanan pelatihan IT.  Kegiatan tersebut dilaksanakan setiap tahun dengan ketentuan anggota organisasi tersebut dapat bergabung ke beberapa tim sebagai ketua atau anggota. Terdapat pula anggota yang sudah tidak aktif di organisasi tersebut (terhapus di database mereka —> label N/A).  Organisasi tersebut juga melakukan rekrutmen volunteer, namun tidak semua tim menggunakan bantuan volunteer. Hasil kegiatan dari setiap tim dapat diwujudkan ke dalam sebuah pengakuan hak cipta atau publikasi karya ilmiah atau publikasi berita di media massa.

## Ketentuan
1. Pengerjaan secara individu alias setiap mahasiswa!
2. Terdapat data dengan nilai #N/A, pertimbangkan cara terbaik untuk mengatasi nilai tersebut.
3. Tidak boleh ada duplikasi pekerjaan antar mahasiswa
4. Berikan penjelasan untuk setiap tahapan. Jangan hanya kode!

## Soal
1. Lakukan pemilahan data training dan testing yang sesuai dan proporsional dari data tersebut! (10 poin)
2. Tentukan data validasi! (10 poin)
3. Tentukan secara tepat kolom yang menjadi "label" (20 poin)
4. Terapkan dua jenis metode klasifikasi dan lakukan perbandingan akurasi  dengan wajib melibatkan variabel keterlibatan volunteer sampai variabel luaran-luaran yang dihasilkan! (20 poin)
5. Lakukan tunning hyperparameter berdasarkan hasil klasifikasi pada nomor 3 menggunakan metode Grid Search dan Random Search! Simpulkan parameter apa yang paling baik digunakan berdasarkan metode klasifikasi yang dipilih, dan berapa akurasinya. (40 poin)

In [87]:
import pandas as pd

df = pd.read_excel('data/dataset_quiz.xlsx')

df

id_ketua  id_anggota1  id_anggota2  id_anggota3  id_anggota4  id_anggota5  \
0     7206.0       7214.0       6211.0       8602.0       7801.0          NaN   
1     7214.0       7109.0       8402.0       7903.0       9202.0          NaN   
2     5911.0       8201.0       8701.0       9101.0       9003.0          NaN   
3     8402.0       7702.0       8602.0       8903.0       8903.0          NaN   
4     8201.0       8305.0       5911.0       5903.0       8402.0          NaN   
..       ...          ...          ...          ...          ...          ...   
80    8803.0       8804.0       9102.0       8005.0       9303.0          NaN   
81    6211.0       8901.0       9003.0       9303.0       9007.0          NaN   
82    8601.0       8903.0       8703.0          NaN          NaN          NaN   
83    8703.0       8601.0       9202.0       9102.0       6204.0          NaN   
84    8102.0       7109.0       5903.0       7206.0       8903.0          NaN   

   keterlibatan_vltr  jml_vltr_terlibat  tahun    skema luaran_hki  \
0              TIDAK                  0   2019  REGULER  TIDAK ADA   
1              TIDAK                  0   2019  REGULER  TIDAK ADA   
2              TIDAK                  0   2019  REGULER  TIDAK ADA   
3              TIDAK                  0   2019  REGULER  TIDAK ADA   
4              TIDAK                  0   2019  REGULER  TIDAK ADA   
..               ...                ...    ...      ...        ...   
80                YA                  3   2022  REGULER  TIDAK ADA   
81                YA                  3   2022  REGULER  TIDAK ADA   
82                YA                  3   2022  REGULER  TIDAK ADA   
83                YA                  3   2022  REGULER  TIDAK ADA   
84             TIDAK                  0   2022  REGULER  TIDAK ADA   

   jenis_luaran_hki potensi_hki luaran_lain jenis_luaran_lain  \
0         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
1         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
2         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
3         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
4         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
..              ...         ...         ...               ...   
80        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
81        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
82        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
83        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
84        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   

   status_luaran_lain  
0           TIDAK ADA  
1           TIDAK ADA  
2           TIDAK ADA  
3           TIDAK ADA  
4           TIDAK ADA  
..                ...  
80          TIDAK ADA  
81          TIDAK ADA  
82          TIDAK ADA  
83          TIDAK ADA  
84          TIDAK ADA  

[85 rows x 16 columns]

# Cleaning Data
Turn data into good data.

In [88]:
# Mengecek nilai yang null

df.isna().sum()

id_ketua               3
id_anggota1            5
id_anggota2            4
id_anggota3            3
id_anggota4           13
id_anggota5           84
keterlibatan_vltr      0
jml_vltr_terlibat      0
tahun                  0
skema                  0
luaran_hki             0
jenis_luaran_hki       0
potensi_hki            0
luaran_lain            0
jenis_luaran_lain      0
status_luaran_lain     0
dtype: int64

### Mengganti Nilai yang hilang / null 

In [89]:
# Mengisi nilai data "id_ketua,id_anggota1,id_anggota2,id_anggota3" dengan menggunakan modus dari tiap" data tersebut

df['id_ketua']  = df['id_ketua'].fillna(df['id_ketua'].mode()[0])
df['id_anggota1']  = df['id_anggota1'].fillna(df['id_anggota1'].mode()[0])
df['id_anggota2']  = df['id_anggota2'].fillna(df['id_anggota2'].mode()[0])
df['id_anggota3']  = df['id_anggota3'].fillna(df['id_anggota3'].mode()[0])

# Mengisi nilai data "id_anggota4" dengan menggunakan mean / rata" dari data tersebut.

df['id_anggota4']  = df['id_anggota4'].fillna(df['id_anggota4'].mean())

# Menghapus nilai data "id_anggota5", karena data tersebut hanya memiliki 1 baris nilai ,
# sehingga menurut saya jika dihapus tidak akan berefek kepada datanya.

df.drop('id_anggota5', axis=1, inplace=True) 

In [90]:
# Mengecek apakah masih ada nilai null setelah mengganti nilai tersebut

df.isna().sum()

id_ketua              0
id_anggota1           0
id_anggota2           0
id_anggota3           0
id_anggota4           0
keterlibatan_vltr     0
jml_vltr_terlibat     0
tahun                 0
skema                 0
luaran_hki            0
jenis_luaran_hki      0
potensi_hki           0
luaran_lain           0
jenis_luaran_lain     0
status_luaran_lain    0
dtype: int64

### Menggunakan LabelEncoder untuk beberapa kolom

In [91]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

cols = ['keterlibatan_vltr', 'skema', 'luaran_hki', 'jenis_luaran_hki', 
        'potensi_hki', 'luaran_lain', 'jenis_luaran_lain', 'status_luaran_lain']

# Encode labels of multiple columns at once

df[cols] = df[cols].apply(le.fit_transform)

In [92]:
df.head(10)

id_ketua  id_anggota1  id_anggota2  id_anggota3  id_anggota4  \
0    7206.0       7214.0       6211.0       8602.0       7801.0   
1    7214.0       7109.0       8402.0       7903.0       9202.0   
2    5911.0       8201.0       8701.0       9101.0       9003.0   
3    8402.0       7702.0       8602.0       8903.0       8903.0   
4    8201.0       8305.0       5911.0       5903.0       8402.0   
5    8008.0       7607.0       7903.0       7702.0       8905.0   
6    8402.0       8803.0       9102.0       8804.0       8005.0   
7    8101.0       7702.0       7308.0       8602.0       9003.0   
8    8402.0       8805.0       8807.0       8803.0       9003.0   
9    8805.0       7206.0       7903.0       7702.0       9002.0   

   keterlibatan_vltr  jml_vltr_terlibat  tahun  skema  luaran_hki  \
0                  0                  0   2019      1           1   
1                  0                  0   2019      1           1   
2                  0                  0   2019      1           1   
3                  0                  0   2019      1           1   
4                  0                  0   2019      1           1   
5                  0                  0   2019      1           1   
6                  0                  0   2019      1           1   
7                  0                  0   2019      1           1   
8                  0                  0   2019      1           1   
9                  0                  0   2019      1           1   

   jenis_luaran_hki  potensi_hki  luaran_lain  jenis_luaran_lain  \
0                 1            2            0                  3   
1                 1            2            0                  3   
2                 1            2            0                  3   
3                 1            2            0                  3   
4                 1            2            0                  3   
5                 1            2            0                  3   
6                 1            2            0                  3   
7                 1            2            0                  3   
8                 1            2            0                  3   
9                 1            2            0                  3   

   status_luaran_lain  
0                   1  
1                   1  
2                   1  
3                   1  
4                   1  
5                   1  
6                   1  
7                   1  
8                   1  
9                   1

### Seleksi fitur

In [93]:
x = df.iloc[:,6:] #Fitur
y = df['keterlibatan_vltr'] #Label
x.shape

(85, 9)

## Melakukan pemilahan data training dan testing yang sesuai dan proporsional.

In [94]:
from sklearn.model_selection import train_test_split

# Split Data Main 
x_main, x_test, y_main, y_test = train_test_split(x, y, test_size=0.2, random_state=7)

# Split Data main menjadi data val dan data train
x_val, x_train, y_val, y_train = train_test_split(x_main, y_main, test_size=0.5, random_state=5)

## Validasi Data

In [95]:
# Split data main dan dan lainnya
# data lainnya, akan kita split lagi menjadi validasi dan train.
df_main, df_test= train_test_split(df, test_size=0.2)

# Split lagi antara validasi dan train
df_val, df_train = train_test_split(df_test, test_size=0.5)

# Cek masing-masing ukuran data

print(f'Jumlah data asli: {df.shape[0]}')
print(f'Jumlah data train: {df_main.shape[0]}')
print(f'Jumlah data val: {df_val.shape[0]}')
print(f'Jumlah data test: {df_train.shape[0]}')

# Cek rasio 'keterlibatan_vltr" label
print('=========')
print(f'Jumlah label data Asli:\n{df.keterlibatan_vltr.value_counts()}')
print(f'Jumlah label data Main:\n{df_main.keterlibatan_vltr.value_counts()}')
print(f'Jumlah label data Val:\n{df_val.keterlibatan_vltr.value_counts()}')
print(f'Jumlah label data Train:\n{df_train.keterlibatan_vltr.value_counts()}')

Jumlah data asli: 85
Jumlah data train: 68
Jumlah data val: 8
Jumlah data test: 9
Jumlah label data Asli:
0    45
1    40
Name: keterlibatan_vltr, dtype: int64
Jumlah label data Main:
0    38
1    30
Name: keterlibatan_vltr, dtype: int64
Jumlah label data Val:
1    5
0    3
Name: keterlibatan_vltr, dtype: int64
Jumlah label data Train:
1    5
0    4
Name: keterlibatan_vltr, dtype: int64


### K-Fold Cross Validation (Random)

In [96]:
# Implementasi k-fold cross validation (random) dengan training, validation, dan testing data
from sklearn.model_selection import train_test_split, KFold

# Split dulu antara data main dan testing dengan train_test_split
df_main, df_test = train_test_split(df, test_size=0.2, random_state=0)

# inisiasi obyek kfold
kf = KFold(n_splits=4)
print(f'Jumlah fold: {kf.get_n_splits()}')
print(f'Obyek KFold: {kf}')
print('--------------------------')

# Lakukan splitting dengan KFold untuk data df_main
# Dengan cara ini, kita masih memiliki data testing untuk keperluan pengujian model
# namun tetap dapat melakukan evaluasi dengan menggunakan data validasi
kf_split = kf.split(df_main)
print(f'Jumlah data df_main: {df_main.shape[0]}')

# cek index data tiap fold
for main_index, test_index in kf_split:
    print(f'Index train: {main_index} | Index test: {test_index}')

Jumlah fold: 4
Obyek KFold: KFold(n_splits=4, random_state=None, shuffle=False)
--------------------------
Jumlah data df_main: 68
Index train: [17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40
 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64
 65 66 67] | Index test: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16]
Index train: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 34 35 36 37 38 39 40
 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64
 65 66 67] | Index test: [17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33]
Index train: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 51 52 53 54 55 56 57 58 59 60 61 62 63 64
 65 66 67] | Index test: [34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50]
Index train: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47

## Kolom yang menjadi Label

In [97]:
y = df['keterlibatan_vltr']
y.head(10)

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
9    0
Name: keterlibatan_vltr, dtype: int32

## Menerapkan 2 jenis metode klasifikasi dengan wajib melibatkan variabel keterlibatan volunteer sampai variabel luaran-luaran

In [98]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier

### Klasifikasi Decision Tree

In [99]:
# Secara default, 
# ecisionTreeClassifier dari scikit-learn akan menggunakan nilai "Gini" untuk kriteria
# Pada kasus ini menggunakan parameter default
dt = DecisionTreeClassifier()

# Sesuaikan dt ke set training
dt.fit(x_main, y_main)

# Memprediksi label set test
y_pred_dt = dt.predict(x_test)

#  menghitung set accuracy
acc_dt = accuracy_score(y_test, y_pred_dt)
print("Test set accuracy: {:.2f}".format(acc_dt))
print(f"Test set accuracy: {acc_dt}")

Test set accuracy: 0.94
Test set accuracy: 0.9411764705882353


### Klasifikasi RandomForest

In [100]:
# RF dapat  digunakan untuk Feature Selection
# Pada kasus kali ini menggunakan estimator pada RandomForest

rf = RandomForestClassifier(n_estimators=10, random_state=1)
#n_estimators = jmlh pohon yang dibuat (split pohon), belum tentu pohon banyak persen lebih banyak
#harus mencari grid search cv

rf.fit(x_main, y_main)

# Memprediksi label set test
y_pred_rf = rf.predict(x_test)

#  menghitung set accuracy
acc_rf = accuracy_score(y_test, y_pred_rf)
print("Test set accuracy: {:.2f}".format(acc_rf))
print(f"Test set accuracy: {acc_rf}")

Test set accuracy: 0.88
Test set accuracy: 0.8823529411764706


## Tunning Hyperparameter

Metode : Grid Search dan Random Search

In [101]:
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import RepeatedKFold

### Repeated K-Fold Cross Validation

In [102]:
# inisiasi repeated k-fold
cv = RepeatedKFold(n_splits=4, n_repeats=3, random_state=42)
# n_repeats digunakan untuk mendefinisikan berapa kali ulangan

### Grid Search - DecisionTree

In [103]:
# Inisiasi Model
dt_gridSearch = DecisionTreeClassifier()

# Definisikan hyperparameter yang akan digunakan
# sklearn menerima dalam bentuk dictionary
# nama hyperparamater HARUS SESUAI dengan dokumentasi sklearn
params_dt = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'splitter' : ["best", "random"],
    'max_depth': list(range(5,11)),
    'min_samples_leaf' : [1,2,3]
}

# inisiasi gridsearch berdasarkan nilai repeated k-fold dan hyperparameter
grid_dt = GridSearchCV(dt_gridSearch, param_grid=params_dt, cv=cv, n_jobs=-1)

%timeit grid_dt.fit(x, y)

score_gs_dt = grid_dt.score(x,y)

print(f'Hasil evaluasi: {score_gs_dt}')
print(f'Konfigurasi hyperparameter: {grid_dt.best_params_}')

5.1 s ± 193 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi: 1.0
Konfigurasi hyperparameter: {'criterion': 'gini', 'max_depth': 5, 'min_samples_leaf': 1, 'splitter': 'best'}


### Grid Search - RandomForest

In [104]:
#Inisiasi Model
rf_gridSearch = RandomForestClassifier()

params_rf = {
    'n_estimators': [1, 5, 10, 50, 100],
    'max_depth':[3,5,10,None],  
    'max_features':['sqrt', 'log2', None],
    'bootstrap' : [True]
}

# inisiasi gridsearch berdasarkan nilai repeated k-fold dan hyperparameter
grid_rf = GridSearchCV(rf_gridSearch, param_grid=params_rf, cv=cv, n_jobs=-1)

%timeit grid_rf.fit(x, y)

score_gs_rf = grid_rf.score(x,y)

print(f'Hasil evaluasi: {score_gs_rf}')
print(f'Konfigurasi hyperparameter: {grid_rf.best_params_}')

30.7 s ± 1.59 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi: 0.9882352941176471
Konfigurasi hyperparameter: {'bootstrap': True, 'max_depth': 3, 'max_features': 'sqrt', 'n_estimators': 100}


### Random Search - DecisionTree

In [105]:
# inisiasi model
dt_randomSearch = DecisionTreeClassifier()

# Definisikan hyperparameter yang akan digunakan
# sklearn menerima dalam bentuk dictionary
# nama hyperparamater HARUS SESUAI dengan dokumentasi sklearn
params_dt = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'splitter' : ["best", "random"],
    'max_depth': list(range(5,11)),
    'min_samples_leaf' : [1,2,3]
}

# inisiasi randomizedsearch berdasarkan nilai repeated k-fold dan hyperparameter
rs_dt = RandomizedSearchCV(dt_randomSearch, param_distributions=params_dt, cv=cv, n_jobs=-1)

# Fit / latih berdasarkan RandomSearch
# %timeit merupakan magic command didalam ipython notebook
# yang dapat kita gunakan untuk menghitung waktu komputasi
# cara ini cukup efektif untuk melakukan evaluasi suatu algoritma atau prosedur
%timeit rs_dt.fit(x, y)

# Evaluasi dengan score
score_rs_dt = rs_dt.score(x,y)

print(f'Hasil evaluasi: {score_rs_dt}')
print(f'Konfigurasi hyperparameter: {rs_dt.best_params_}')

608 ms ± 26.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi: 1.0
Konfigurasi hyperparameter: {'splitter': 'best', 'min_samples_leaf': 1, 'max_depth': 9, 'criterion': 'gini'}


## RandomSearch - RandomForest

In [106]:
#Inisiasi Model
rf_randomSearch = RandomForestClassifier()

params_rf = {
    'n_estimators': [1, 5, 10, 50, 100],
    'max_depth':[3,5,10,None],  
    'max_features':['sqrt', 'log2', None],
    'bootstrap' : [True]
}

# inisiasi gridsearch berdasarkan nilai repeated k-fold dan hyperparameter
rs_rf = RandomizedSearchCV(rf_randomSearch, param_distributions=params_rf, cv=cv, n_jobs=-1)

%timeit rs_rf.fit(x, y)

score_rs_rf = rs_rf.score(x,y)

print(f'Hasil evaluasi: {score_rs_rf}')
print(f'Konfigurasi hyperparameter: {rs_rf.best_params_}')

5.15 s ± 1.07 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi: 1.0
Konfigurasi hyperparameter: {'n_estimators': 100, 'max_features': None, 'max_depth': 3, 'bootstrap': True}


## Accuracy Score :  
Hasil accuracy score sebelum dan setelah dilakukan tunning hyperparameter :

- Decision Tree  
0.9411764705882353 => 1.0 (Tunning Hyperparameter GridSearch)  
0.9411764705882353 => 1.0 (Tunning Hyperparameter RandomSearch)  
- RandomForest  
0.8823529411764706 => 0.9882352941176471 (Tunning Hyperparameter GridSearch)  
0.8823529411764706 => 1.0 (Tunning Hyperparameter RandomSearch)